From 04b74ed2d526efa4f6dcc526696f73847d04c168 Mon Sep 17 00:00:00 2001
From: Daniel Veillard <veillard@redhat.com>
Date: Fri, 26 Oct 2012 13:50:47 +0800
Subject: [PATCH] Fix large parse of file from memory
To: libvir-list@redhat.com

https://bugzilla.redhat.com/show_bug.cgi?id=862969
The new code trying to detect excessive input lookup would
just get wrong sometimes in the case of very large file parsed
directly from memory.

Signed-off-by: Daniel Veillard <veillard@redhat.com>
---
 libxml.h | 2 ++
 parser.c | 1 +
 xmlIO.c  | 2 +-
 3 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/libxml.h b/libxml.h
index 3c44c83..6f43f7b 100644
--- a/libxml.h
+++ b/libxml.h
@@ -79,6 +79,8 @@ void __xmlGlobalInitMutexLock(void);
 void __xmlGlobalInitMutexUnlock(void);
 void __xmlGlobalInitMutexDestroy(void);
 
+int xmlNop(void);
+
 #ifdef IN_LIBXML
 #ifdef __GNUC__
 #ifdef PIC
diff --git a/parser.c b/parser.c
index 856eb8b..3138098 100644
--- a/parser.c
+++ b/parser.c
@@ -1913,6 +1913,7 @@ static void xmlSHRINK (xmlParserCtxtPtr ctxt) {
 static void xmlGROW (xmlParserCtxtPtr ctxt) {
     if ((((ctxt->input->end - ctxt->input->cur) > XML_MAX_LOOKUP_LIMIT) ||
          ((ctxt->input->cur - ctxt->input->base) > XML_MAX_LOOKUP_LIMIT)) &&
+         ((ctxt->input->buf) && (ctxt->input->buf->readcallback != xmlNop)) &&
         ((ctxt->options & XML_PARSE_HUGE) == 0)) {
         xmlFatalErr(ctxt, XML_ERR_INTERNAL_ERROR, "Huge input lookup");
         xmlHaltParser(ctxt);
diff --git a/xmlIO.c b/xmlIO.c
index 8fc00e3..dc9f410 100644
--- a/xmlIO.c
+++ b/xmlIO.c
@@ -786,7 +786,7 @@ xmlCheckFilename (const char *path)
     return 1;
 }
 
-static int
+int
 xmlNop(void) {
     return(0);
 }
-- 
2.5.0

